Basic support for clang 10 with X86

nashif opened the pull request:
I can build with clang 10, I can also run some of the tests and samples..
Userspace currently fails, for example:

`west build -b qemu_x86_64 tests/kernel/semaphore/semaphore  -t run`

results in:

```

*** Booting Zephyr OS build zephyr-v2.2.0-2235-g219cc85aa116  ***
Running test suite test_semaphore
===================================================================
starting test - test_k_sema_init
PASS - test_k_sema_init
===================================================================
starting test - test_sema_thread2thread
E: 0x000000000017e000 is not a valid z_thread_stack_element
E: syscall z_vrfy_k_thread_create failed check: bad stack object
E: Bad system call from RIP 0x100629
E: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0
E: Current thread: 0x000000000012cf20 (unknown)
E: Halting system
FAILED: zephyr/CMakeFiles/run
```

`west build -b qemu_x86 samples/philosophers/  -t run` works fine.
`west build -b qemu_x86_64 samples/philosophers/  -t run` does not...

nashif committed:
toolchain: make clang happy with inline asm

clang did not like %p on x86, make it use %c.

Did not explore why this is happening..

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
clang: do not assert on libgcc not found

We do not need libgcc always, some environments do not have libgcc and
do not require it, so keep it more flexible.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
This works with the installed clang/llvm installed on fedora 31, only set toolchain variant to llvm.

nashif commented:
llvm 10 was complain on Mac OS about the -T option while linking, need to see what is going on there...

andrewboie commented:
> starting test - test_sema_thread2thread
> E: 0x000000000017e000 is not a valid z_thread_stack_element
> E: syscall z_vrfy_k_thread_create failed check: bad stack object
> E: Bad system call from RIP 0x100629
> E: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0
> E: Current thread: 0x000000000012cf20 (unknown)
> E: Halting system

Might be an issue with data addresses shifting between zephyr_prebuilt.elf and zephyr.elf?

ceolin commented:
In Fedora 31 I'm getting this:

```
Including boilerplate (Zephyr base): /home/faceolin/p/zephyr/cmake/app/boilerplate.cmake
-- Application: /home/faceolin/p/zephyr/tests/kernel/semaphore/semaphore
-- Zephyr version: 2.2.99 (/home/faceolin/p/zephyr)
-- Found Python3: /usr/bin/python3.7 (found suitable exact version "3.7.6") found components: Interpreter 
-- Board: qemu_x86_64
-- Found west: /home/faceolin/.local/bin/west (found suitable version "0.7.2", minimum required is "0.7.1")
-- Found toolchain: host (clang/ld)
-- Found BOARD.dts: /home/faceolin/p/zephyr/boards/x86/qemu_x86/qemu_x86_64.dts
-- Generated zephyr.dts: /home/faceolin/p/build/zephyr/zephyr.dts
-- Generated devicetree_unfixed.h: /home/faceolin/p/build/zephyr/include/generated/devicetree_unfixed.h
Parsing /home/faceolin/p/zephyr/Kconfig
Loaded configuration '/home/faceolin/p/zephyr/boards/x86/qemu_x86/qemu_x86_64_defconfig'
Merged configuration '/home/faceolin/p/zephyr/tests/kernel/semaphore/semaphore/prj.conf'
Configuration saved to '/home/faceolin/p/build/zephyr/.config'
Kconfig header saved to '/home/faceolin/p/build/zephyr/include/generated/autoconf.h'
-- The C compiler identification is Clang 9.0.0
-- The CXX compiler identification is Clang 9.0.0
-- The ASM compiler identification is Clang
-- Found assembler: /usr/bin/clang
-- Cache files will be written to: /home/faceolin/.cache/zephyr
CMake Error at ../../../../CMakeLists.txt:250 (toolchain_ld_base):
  Unknown CMake command "toolchain_ld_base".


-- Configuring incomplete, errors occurred!
See also "/home/faceolin/p/build/CMakeFiles/CMakeOutput.log".
See also "/home/faceolin/p/build/CMakeFiles/CMakeError.log".
```

`toolchain_ld_base` macro is defined in `cmake/linker/ld/target`, shouldn't we have something equivalent to `ldd` now ? /me guessing

ceolin commented:
> In Fedora 31 I'm getting this:
> 
> ```
> Including boilerplate (Zephyr base): /home/faceolin/p/zephyr/cmake/app/boilerplate.cmake
> -- Application: /home/faceolin/p/zephyr/tests/kernel/semaphore/semaphore
> -- Zephyr version: 2.2.99 (/home/faceolin/p/zephyr)
> -- Found Python3: /usr/bin/python3.7 (found suitable exact version "3.7.6") found components: Interpreter 
> -- Board: qemu_x86_64
> -- Found west: /home/faceolin/.local/bin/west (found suitable version "0.7.2", minimum required is "0.7.1")
> -- Found toolchain: host (clang/ld)
> -- Found BOARD.dts: /home/faceolin/p/zephyr/boards/x86/qemu_x86/qemu_x86_64.dts
> -- Generated zephyr.dts: /home/faceolin/p/build/zephyr/zephyr.dts
> -- Generated devicetree_unfixed.h: /home/faceolin/p/build/zephyr/include/generated/devicetree_unfixed.h
> Parsing /home/faceolin/p/zephyr/Kconfig
> Loaded configuration '/home/faceolin/p/zephyr/boards/x86/qemu_x86/qemu_x86_64_defconfig'
> Merged configuration '/home/faceolin/p/zephyr/tests/kernel/semaphore/semaphore/prj.conf'
> Configuration saved to '/home/faceolin/p/build/zephyr/.config'
> Kconfig header saved to '/home/faceolin/p/build/zephyr/include/generated/autoconf.h'
> -- The C compiler identification is Clang 9.0.0
> -- The CXX compiler identification is Clang 9.0.0
> -- The ASM compiler identification is Clang
> -- Found assembler: /usr/bin/clang
> -- Cache files will be written to: /home/faceolin/.cache/zephyr
> CMake Error at ../../../../CMakeLists.txt:250 (toolchain_ld_base):
>   Unknown CMake command "toolchain_ld_base".
> 
> 
> -- Configuring incomplete, errors occurred!
> See also "/home/faceolin/p/build/CMakeFiles/CMakeOutput.log".
> See also "/home/faceolin/p/build/CMakeFiles/CMakeError.log".
> ```
> 
> `toolchain_ld_base` macro is defined in `cmake/linker/ld/target`, shouldn't we have something equivalent to `ldd` now ? /me guessing

Just confirming that I was able to build using `ld`, don't know what is the difference in our environment. That's being said, I also could use clang6 with this pr, same ld issue though

nashif commented:
> Just confirming that I was able to build using `ld`, don't know what is the difference in our environment. That's being said, I also could use clang6 with this pr, same ld issue though

yeah, somewhere I set the linker to be ldd instead of ld and did not including the required changes in this PR.

```
-set(LINKER ld) # TODO: Use lld eventually rather than GNU ld
+set(LINKER lld)
```

Need to move to ldd, not sure however why it did not complain on my side.

nashif committed:
boards: qemu_x86: enable llvm toolchain

Make this board buildable with llvm.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

ceolin commented:
Tried again with clang-10 this time and got same results as you.

tejlmand commented:
@nashif what's the plan on this ?
It works with clang 10, but I notice an assert has been commented out.
Same with linker flags.

I assume to get it working with clang 10.

tejlmand commented:
Works on Ubuntu 18.04:
```
 west build -b qemu_x86_64 samples/philosophers/ -t run
qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.x2apic [bit 21]
SeaBIOS (version rel-1.12.1-0-ga5cab58-dirty-20200516_152813-e7ee184eb100-zephyr
)
Booting from ROM..Optimal CONFIG_X86_MMU_PAGE_POOL_PAGES 13
*** Booting Zephyr OS build zephyr-v2.3.0-328-g81be72259037  ***
Running test suite test_semaphore
===================================================================
starting test - test_k_sema_init
PASS - test_k_sema_init
===================================================================
starting test - test_sema_thread2thread
PASS - test_sema_thread2thread
===================================================================
starting test - test_sema_thread2isr
PASS - test_sema_thread2isr
===================================================================
starting test - test_sema_reset
PASS - test_sema_reset
===================================================================
starting test - test_sema_count_get
PASS - test_sema_count_get
===================================================================
starting test - test_simple_sem_from_isr
PASS - test_simple_sem_from_isr
===================================================================
starting test - test_simple_sem_from_task
PASS - test_simple_sem_from_task
===================================================================
starting test - test_sem_take_no_wait
PASS - test_sem_take_no_wait
===================================================================
starting test - test_sem_take_no_wait_fails
PASS - test_sem_take_no_wait_fails
===================================================================
starting test - test_sem_take_timeout_fails
PASS - test_sem_take_timeout_fails
===================================================================
starting test - test_sem_take_timeout
PASS - test_sem_take_timeout
===================================================================
starting test - test_sem_take_timeout_forever
PASS - test_sem_take_timeout_forever
===================================================================
starting test - test_sem_take_timeout_isr
PASS - test_sem_take_timeout_isr
===================================================================
starting test - test_sem_take_multiple
PASS - test_sem_take_multiple
===================================================================
starting test - test_sem_give_take_from_isr
PASS - test_sem_give_take_from_isr
===================================================================
starting test - test_sem_multiple_threads_wait
PASS - test_sem_multiple_threads_wait
===================================================================
starting test - test_sem_measure_timeouts
PASS - test_sem_measure_timeouts
===================================================================
starting test - test_sem_measure_timeout_from_thread
PASS - test_sem_measure_timeout_from_thread
===================================================================
starting test - test_sem_multiple_take_and_timeouts
PASS - test_sem_multiple_take_and_timeouts
===================================================================
starting test - test_sem_multi_take_timeout_diff_sem
PASS - test_sem_multi_take_timeout_diff_sem
===================================================================
Test suite test_semaphore succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
qemu-system-x86_64: terminating on signal 2
```

nashif committed:
toolchain: clang: add -Wno-typedef-redefinition option

Without the -Wno-typedef-redefinition option, clang complains if a
typedef gets redefined in gnu99  mode (since this is officially a C11
feature).

While new versions of GCC do not seem to issue this warning in gnu99
mode anymore. So some existing code with typedef redefined which works
well with GCC will issue this warning.

Similar to what was done in 2354f055ec41fa49f8fde5f5ffca0f84313b7544.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
@tejlmand this should be considered now, please review.

tejlmand reviewed (changes requested):
Some small comments.

Also I get a warning from llvm-objdump.
```
/opt/llvm/clang+llvm-10.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/llvm-objdump: warning: 'zephyr.elf': failed to parse debug information for zephyr.elf
```
but unfortunately I don't think we can do much on that.
http://llvm.1065342.n5.nabble.com/llvm-dev-llvm-objdump-failed-to-parse-debug-information-td137299.html
https://bugs.llvm.org/show_bug.cgi?id=41905
https://reviews.llvm.org/D62462

tejlmand commented:
One more observation, which is not directly related to llvm (and thus this PR), but that we must take care of, is the use of `LOG2CEIL` in linker scripts.
Apparently llvm-ld doesn't support that linker function.

So users linking for nrf52840 (and probably other arm systems) will see this:
```
ld.lld: error: zephyr/linker.cmd:109: , expected, but got (
>>> . = ALIGN( 1 << LOG2CEIL(4 * 32) );
>>>                         ^
clang-10: error: ld.lld command failed with exit code 1 (use -v to see invocation)
```

tejlmand commented:
@evgeniy-paltsev @abrodkin knowing arcmwdt is clang based, could you be in risk of similar issue regarding the `LOG2CEIL` mentioned here: https://github.com/zephyrproject-rtos/zephyr/pull/24917#issuecomment-717916438, if `CONFIG_ARC_MPU_ENABLE=y` ?
https://github.com/zephyrproject-rtos/zephyr/blob/74b0b4116710f9176e027fc9196eca8a3c5a7849/include/arch/arc/v2/linker.ld#L30-L40

nashif committed:
clang: remove include-fixed inclusion

This directory does not exist in clang/llvm, remove it.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
clang: remove cache handling

This is taking too long for every run and does not seem to work as
expected and wastes time instead of saving it. Remove for now..

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
llvm: add support for building x86_64-pc

set triplet for 64bit x86.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
clang: use LLVM_TOOLCHAIN_PATH instead of CLANG_ROOT_DIR

Be consistent with other toolchains.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

tejlmand commented on cmake/toolchain/llvm/generic.cmake:
```diff
@@ -20,7 +16,11 @@ if("${ARCH}" STREQUAL "arm")
   set(triple arm-none-eabi)
```
We could consider if it would be beneficial to use the:
https://clang.llvm.org/docs/CrossCompilation.html#target-triple
sub for arm, like: `v6m`, `v7m`, `v7a`, ...

although impact might be limited as we already are setting `-mcpu`.

tejlmand reviewed (approved):
Look good

evgeniy-paltsev commented:
> @evgeniy-paltsev @abrodkin knowing arcmwdt is clang based, could you be in risk of similar issue regarding the `LOG2CEIL` mentioned here: [#24917 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/24917#issuecomment-717916438), if `CONFIG_ARC_MPU_ENABLE=y` ?
> https://github.com/zephyrproject-rtos/zephyr/blob/74b0b4116710f9176e027fc9196eca8a3c5a7849/include/arch/arc/v2/linker.ld#L30-L40
> 
> or does the arcmwdt linker support `LOG2CEIL` ?

We support it via macros, check it here:
https://github.com/zephyrproject-rtos/zephyr/blob/74b0b4116710f9176e027fc9196eca8a3c5a7849/include/linker/linker-tool-mwdt.h#L24

The ARC MWDT linker doesn't support `LOG2CEIL`.
